### PR TITLE
Multi instance push refresh

### PR DESCRIPTION
**Pull request by mrm9084**
Adds the ability to trigger refresh of configuration via a push notification from event grid for applications with multiple instances using Service Bus. Works similar to Spring Cloud Bus Refresh.

### CODE REVIEW

**Review by saragluna — COMMENTED** [review]
So when you have several application instances in the cluster, and call the refresh endpoint of one of them, the refresh will then be published to all other instances? Which module/starter of Service Bus will be added when using this `push refresh`?

**Comment by mrm9084** [comment]
> So when you have several application instances in the cluster, and call the refresh endpoint of one of them, the refresh will then be published to all other instances? Which module/starter of Service Bus will be added when using this `push refresh`?

@saragluna For multiple instances you need to setup; spring-cloud-bus, azure-spring-cloud-stream-binder-servicebus-topic.

You then need each of your instance to have their own subscription and call the appconfiguration-refresh-bus endpoint. I am working on a ms.docs doc for it.

File: sdk/appconfiguration/azure-spring-cloud-appconfiguration-config-web/src/main/java/com/microsoft/azure/spring/cloud/config/web/pushbusrefresh/AppConfigurationBusRefreshEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    public AppConfigurationBusRefreshEndpoint(ApplicationEventPublisher context, String appId,
+            AppConfigurationProperties appConfiguration) {
+        super(context, appId);
+        this.appConfiguration = appConfiguration;
+    }
+
+    @PostMapping(value = "/")
+    @ResponseBody
+    public String refresh(HttpServletRequest request, HttpServletResponse response,
```
Comment (backwind1233) [inline sdk/appconfiguration/azure-spring-cloud-appconfiguration-config-web/src/main/java/com/microsoft/azure/spring/cloud/config/web/pushbusrefresh/AppConfigurationBusRefreshEndpoint.java:51]: Hi, @mrm9084  

>Multi instance push refresh

I don't know what the "multi" means, could you explain it.

**Comment by backwind1233** [comment]
> > So when you have several application instances in the cluster, and call the refresh endpoint of one of them, the refresh will then be published to all other instances? Which module/starter of Service Bus will be added when using this `push refresh`?
> 
> @saragluna For multiple instances you need to setup; spring-cloud-bus, azure-spring-cloud-stream-binder-servicebus-topic.
> 
> You then need each of your instance to have their own subscription and call the appconfiguration-refresh-bus endpoint. I am working on a ms.docs doc for it.

hi, @mrm9084  if there are 3 nodes of the same application, and one of them call the refresh endpoint, will the other 2 nodes still need call the refresh endpoint?

**Comment by mrm9084** [comment]
> > > So when you have several application instances in the cluster, and call the refresh endpoint of one of them, the refresh will then be published to all other instances? Which module/starter of Service Bus will be added when using this `push refresh`?
> > 
> > 
> > @saragluna For multiple instances you need to setup; spring-cloud-bus, azure-spring-cloud-stream-binder-servicebus-topic.
> > You then need each of your instance to have their own subscription and call the appconfiguration-refresh-bus endpoint. I am working on a ms.docs doc for it.
> 
> hi, @mrm9084 if there are 3 nodes of the same application, and one of them call the refresh endpoint, will the other 2 nodes still need call the refresh endpoint?

@backwind I think you missunderstand. The original all to the endpoint is done via Event Hub which can be configured to call a Web Hook. If the Endpoint is the appconfiguration-refresh-bus, and everything is setup then the client called will publish a messaged to Service Bus, or really any message broker, I have tested with RabbitMQ also. The other clients are all subscribed to the Service Bus get the notification.

It works the same way as the Service Bus Topic example.

**Review by saragluna — APPROVED** [review]
LGTM.

**Pull request merged**